### PR TITLE
Ensure that `-transpose` also transposes the zoom order (needed for anisotropic images)

### LINKED ISSUE #4417: `sct_image` option `-transpose` doesn't transpose zooms (incorrect for anisotropic images)
While working on https://forum.spinalcordmri.org/t/sct-deepseg-mouse-sc-segmentation-error/1237, I needed to fix the user's data using `-transpose` (which transposes the data array but not the header). But, I found that the results were stretched and skewed.

Before (`vol0001.nii`):

image

After (`sct_image -i vol0001.nii -transpose x,z,y -o vol0001_transposed.nii`):

image

The issue here is that the data array was transposed, but the zooms were not. Meaning, for this anisotropic image, the wrong scaling is being applied to each axis.

After fixing the problem then re-running the command, the expected result is:

image

Since I implemented the fix to test the problem itself, this should be an easy PR to open.

### PR DESCRIPTION
**Pull request by joshuacwnewton**
## Description

This PR ensures that, when transposing just the data array, we also transpose the zooms (to ensure that the correct resolution is applied for anisotropic images).

Issue #4417 contains an illustrative comparison for when A) zooms aren't transposed and B) zooms are correctly transposed.

## Alternative solution

While writing this simple fix, I also noticed that we perform similar transposing during the `-setorient` command:

- `-setorient`: Change the axis order for both the data array _and_ the header.
- `-transpose`: Change the axis order for _just_ the data array.

However, `-setorient` [isn't affected by this bug](https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4417#issuecomment-2040175904)! So, I was wondering if it might be better to perform a refactor where both `-setorient` and `-transpose` use the same function, but `-transpose` is run with some sort of `data_only=True` argument. That way, we're consistently using the same transpose logic for both arguments.

## Linked issues

Fixes #4417.

### CODE REVIEW

**Comment by mguaypaq** [comment]
Looking at the original forum issue, I'm noticing another wrinkle: I believe the axis naming conventions are different between human MRI images and mouse MRI images. Specifically, for humans the spinal cord goes along the SI direction (because it goes down from the brain when standing upright), whereas in mice I think the spinal cord goes in the AP direction (because it extends behind the brain when standing on four legs). If this is true (I'm really not sure), then I'm not sure the forum user's images are really transposed, but also I'm not sure what orientation assumptions SCT's mouse segmentation model makes.

File: spinalcordtoolbox/scripts/sct_image.py [hunk 1]
```diff
@@ -369,6 +369,7 @@ def main(argv: Sequence[str]):
         transpose_numerical = [DIM_LIST.index(axis)  # Convert [x, y, z, t] to [0, 1, 2, 3]
                                for axis in arguments.transpose]
         im_out.data = np.transpose(im_out.data, axes=transpose_numerical)
+        im_out.header.set_zooms([im_out.header.get_zooms()[ax] for ax in transpose_numerical])
```
[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_image.py:372]: For me, `-transpose` is used purely to ensure that the orientation information in the header matches the orientation of the data array.

Here is a brief example:

- Header says RPI, but data array is oriented as RIP (i.e. x=RL axis, y=IS axis, z=PA axis).
- Two options to fix this:
   1. Change the header to say, e.g. RIP, then use `-setorient` to reorient both the header _and_ the data array back to RPI.
   2. Tranpose the data array to be RPI (i.e. x=RL axis, y=PA axis, z=IS axis).

IMO these two operations should result in the same outcome? So, I would be curious to try that here to see if `-setorient` results in the same mismatched affines, or if both matrixes are changed during the reorientation process.

If both are changed, then maybe that hints that `-transpose` should also change both? :thinking: 

But, I think this also goes back to my earlier comment:

> However, -setorient https://github.com/spinalcordtoolbox/spinalcordtoolbox/issues/4417#issuecomment-2040175904! So, I was wondering if it might be better to perform a refactor where both -setorient and -transpose use the same function, but -transpose is run with some sort of data_only=True argument. That way, we're consistently using the same transpose logic for both arguments.

**Comment by joshuacwnewton** [comment]
> I believe the axis naming conventions are different between human MRI images and mouse MRI images. Specifically, for humans the spinal cord goes along the SI direction (because it goes down from the brain when standing upright), whereas in mice I think the spinal cord goes in the AP direction (because it extends behind the brain when standing on four legs). If this is true (I'm really not sure), then I'm not sure the forum user's images are really transposed, but also I'm not sure what orientation assumptions SCT's mouse segmentation model makes.

!!!

That might explain why the performance of the model was _much_ better when the axes were quote-unquote transposed, and why the performance dropped so suddenly after I "fixed" the data. Very good observation!

File: spinalcordtoolbox/scripts/sct_image.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mguaypaq) [reply spinalcordtoolbox/scripts/sct_image.py:372]: Thanks for the explanation, it clarifies things for me. But actually it's convincing me that `-transpose` should _not_ touch the zooms! Here's why:

* The sform in the header describes not just the orientation (for example, RPI), but also the origin of the coordinate system, and physical dimensions of the voxels along each axis. Essentially, it says "voxel `[0, 0, 0]` is at physical coordinates $(a, b, c)$" and "voxel `[0, 0, 1]` is at physical coordinates $(a', b', c')$", etc. By looking at which the differences $a-a'$ and $b-b'$ and $c-c'$ is the biggest, we can tell whether the z-axis is AP or SI or RL. By looking at the sign of this biggest difference, we can tell whether it's AP/SI/RL or PA/IS/LR. And by looking at the distance $\sqrt{(a-a')^2 + (b-b')^2 + (c-c')^2}$, we get the zoom for the z-axis. So, all of this information is linked.
* If we were to change the zooms, that would mean non-uniformly stretching the sampling grid in physical space, which seems like a big change to me.
* Compared to this, I can easily imagine software that inadvertently transposes the data array, without touching the sform at all.
* In other words, if we assume that the sform has the correct information for _some_ ordering, then we should assume that the zooms are correct for that ordering.
* Meanwhile, the data array is completely independent of any physical dimensions. So, for example, if the data array is somehow in RIP order, the correct way to line this up with an RPI header would be to `np.transpose` the array, _without_ touching the zooms.

[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_image.py:372]: I think there might be something I'm fundamentally misunderstanding about the affine matrix + zooms. :sweat:

My (very naive) understanding was this:

- If `get_zooms()` returns `[2.0, 3.0, 4.0]`, then that means the `x` axis has resolution 2.0mm, the `y` axis has resolution 3.0mm, and the `z` axis has resolution 4.0mm.
- If we transpose the data array by e.g. swapping the y and z axes (but leave the zooms untouched), then now the zooms will refer to the wrong axes (3.0mm will be applied to what was the `z` axis, and 4.0mm will be applied to what was the `y` axis).
- But, from what I understand, this is only actually true for an affine where the diagonal of the inner 3x3 matrix precisely matches the zoom values? As soon as the non-diagonal elements of the inner 3x3 matrix are nonzero, then the mapping from voxel to physical coordinates gets much more complicated?

Outside of the "naive" case, I feel like I've gotten myself a bit confused, and now I'm no longer certain that I've handled the user's data correctly from the previous issue. :sweat:

[the remark this answers is not in the document]
Comment (mguaypaq) [reply spinalcordtoolbox/scripts/sct_image.py:372]: That's ok, it's a confusing topic. My understanding is that:
* `get_zooms()` returns the vector norms for the first three columns of the qform matrix (that is, if the first column is `[x, y, z, w]`, then `get_zooms()[0]` is $\sqrt{x^2+y^2+z^2+w^2}$).
* `set_zooms()` rescales the each of the first three columns of the qform matrix by a factor of `new_zoom[i] / old_zoom[i]`.

In particular, all of these operations leave the sform untouched.

One theory: if the sform and the qform match, then we should assume that only the data array needs to be transposed. But if the sform and the qform are different, then _maybe_ the zooms should be adjusted. But actually, maybe these operations would be enough to "fix" the image:
1. Figure out whether the sform or the qform is correct.
2. Set one to the other, with `sct_image -set-sform-to-qform` or `sct_image -set-qform-to-sform`, so that the header orientation is correct.
3. Figure out the data array's orientation.
4. Use `sct_image -transpose` and/or `sct_image -flip` to fix the data array's orientation so that it matches the header, without touching the zooms.
5. Optionally, do `sct_image -setorient` at the end to get the desired orientation.

If the sform and the qform are both wrong... maybe fixing the image is beyond SCT's mandate.

[the remark this answers is not in the document]
Comment (joshuacwnewton) [reply spinalcordtoolbox/scripts/sct_image.py:372]: Thank you so much for writing out a little protocol for handling data issues.

In that case, perhaps I should close this PR, and instead create a dev wiki page documenting how to troubleshoot data issues, incorporating the steps you've outlined?

This has been quite illuminating for me, and I like the thought of focusing more on my troubleshooting approach than kludging up the tools we use to solve the problems. :)

I have one other question that I'm a bit confused about (and you may not know the answer, that's okay!):

- In both the [`t2.nii.gz` example image](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4434#discussion_r1566297439) you pulled, as well as the user's data, the 3x3 matrix is non-diagonal, but in a peculiar, special-case sort of way: most of the cells will be `0.0`, with one cell in each column being nonzero, and sometimes negative? Almost like a typical diagonal matrix, but with the values shuffled about.
- Based on my basic understanding of affine matrices: diagonal == scaling, while non-diagonal == skewing/rotation/etc.
- Assuming the goal is simply to scale in most cases, why on earth are some of these affine matrices non-diagonal, then? I can never quite tell if the affine matrix has issues, or if this special-case non-diagonal matrix has a specific (useful) effect.

[the remark this answers is not in the document]
Comment (mguaypaq) [reply spinalcordtoolbox/scripts/sct_image.py:372]: Sure, documenting this on the dev wiki sounds like a good idea.

I'm pretty handy with matrices, so let me try an explanation :smile:

Let $x, y, z$ be the physical coordinates (floating point numbers in millimeters, for example). Let $i, j, k$ be the array coordinates of a voxel (so, integers starting at 0). Let $M$ be the first three rows of an sform (so, a 3 by 4 matrix that looks like

$$ M = \\pmatrix{M_{11} & M_{12} & M_{13} & M_{14} \\\\ M_{21} & M_{22} & M_{23} & M_{24} \\\\ M_{31} & M_{32} & M_{33} & M_{34}} $$

Then the formula for converting between array and physical coordinates is:

$$ \\pmatrix{x \\\\ y \\\\ z} = M \\pmatrix{i \\\\ j \\\\ k \\\\ 1} $$

In other words:

$$ x = M_{11} \cdot i + M_{12} \cdot j + M_{13} \cdot k + M_{14} $$

$$ y = M_{21} \cdot i + M_{22} \cdot j + M_{23} \cdot k + M_{24} $$

$$ z = M_{31} \cdot i + M_{32} \cdot j + M_{33} \cdot k + M_{34} $$

From this, we see that the last column of $M$ is actually the physical coordinates of the (0, 0, 0) voxel.

Also, suppose the first three columns are in the special form you mentioned (mostly zeros, with a single nonzero in each row and each column). In fact, let's take the example from my earlier `set_zooms` experiment: $M_{13} = 4.0$, $M_{21} = -2.0$, $M_{32} = 3.0$. Then, the equations become very simple, because only the nonzero terms contribute:

$$ x = 4.0 \cdot k + M_{14} $$

$$ y = -2.0 \cdot i + M_{24} $$

$$ z = 3.0 \cdot j + M_{34} $$

So, in that case, each array coordinate $i, j, k$ controls a single physical coordinate $x, y, z$, but **they can be swapped**. The nonzero entries tell you the mapping:
* The column number of the nonzero entry tells you which of $i, j, k$.
* The row number of the nonzero entry tells you which of $x, y, z$.
* The (absolute) value of the nonzero entry tells you the voxel dimensions along that axis: how many millimeters for each step in the data array.
* The sign of the nonzero entry distinguishes between AP/SI/RL and PA/IS/LR, since it tells us whether the coordinates are increasing or decreasing.

**Comment by jcohenadad** [comment]
Excellent point @mguaypaq https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4434#issuecomment-2057635322. It is true that there are _some_ differences in the descriptive anatomy between human and other vertebrae. As you rightfully pointed out, A-P in quadrupeds respectively refer to rostral-caudal. Useful references:
- https://nba.uth.tmc.edu/neuroscience/m/s2/chapter01.html (scroll down to "1.6 Orientation to the Central Nervous System")
- https://en.wikipedia.org/wiki/Anatomical_terms_of_location

**Comment by joshuacwnewton** [comment]
Closing this PR due as we've concluded that `-tranpose` works as it should (transposing just the data array, leaving the header alone), and [other means](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/4434#discussion_r1567800751) should be used to correct the image.
